### ISSUE TITLE
compare keys by fingerprints whenever possible. Store key fingerprints in db

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-ios#248]
We should start saving fingerprints in the storage (in addition to longids), and compare by fingerprints wherever possible instead of longids. It's safer - collisions less likely. I'll make another issue for this.

_Originally posted by @tomholub in https://github.com/FlowCrypt/flowcrypt-ios/pull/245#discussion_r458311381_

**Comment by tomholub** [comment]
This will involve a database migration

**Comment by tomholub** [comment]
Since you are removing migrations and anyway we break compatibility with previous versions, might as well do this.

**Comment by tomholub** [comment]
We could have @ekievsky look at this after #314

**Comment by Kharchevskyi** [comment]
@ekievsky is it possible to postpone this task until #197 will be resolved? (_probably will have conflicts_)

**Comment by tomholub** [comment]
In that case Evgenii could do https://github.com/FlowCrypt/flowcrypt-ios/issues/296 first

**Comment by ekievsky** [comment]
Right, I am working on #296 right now.

**Comment by ekievsky** [comment]
@tomholub should I checkout from feature/issue-197-pass-phrase-memory or work on other task?

**Comment by Kharchevskyi** [comment]
@ekievsky is it possible to work on other issue? I still need a bit of time to fix this. Otherwise I think it will lead to merge conflicts

**Comment by Kharchevskyi** [comment]
@tomholub 
1. Should this "FingerprintObject" also have longid?
2. Should KeyId use "FingerprintObject"?  
3. Should KeyDetails also have an array of "FingerprintObject"?

For now KeyDetails are compared by fingerprints
```
extension KeyDetails: Equatable {
    static func == (lhs: KeyDetails, rhs: KeyDetails) -> Bool {
        lhs.ids == rhs.ids
    }
}

extension KeyId: Equatable {
    static func == (lhs: KeyId, rhs: KeyId) -> Bool {
        lhs.fingerprint == rhs.fingerprint
    }
}
```

Can you please elaborate on this issue?

**Comment by tomholub** [comment]
I'm not sure what you mean by FingeprintObject.

Cor code (as opposed to storage), right now we have KeyId object which is fine - it just needs `shortid` and `keywords` removed, so that it only has `fingerprint` and `longid`. KeyDetails already uses this, no problem there.

The main goal here is to address storage (`KeyInfo` for Realm) which currently has only `longid`. There are several problems with this:

1) using fingerprint instead of longid is safer / preferred whenever possible
2) openpgp keys have several fingerprint/longid pairs and we should be able to query any of them.

Therefore we should find some way to represent something like `[KeyId]`, modified as mentioned above, in storage along with each key.

But anytime I interact with the key storage, I should be able to run a query like `get a key by fingerprint XXX`, or, in rare situation when I only have longid, `get a key by longid YYY`. It would be even better if you could differentiate first KeyId from the other KeyIds on a particular key - the first one is primary which has a special meaning in OpenPGP. So I could run a query like `get a key where primary fingerprint equals XXX` or I could say `get a key where any fingerprint equals XXX` or `get a key where longid ...` etc.

Right now we only use longid in storage, and only one (the primary), which is both wrong.

Exact same thing applies to public key storage - I should be able to query keys by all of the ways I mentioned above, therefore we also have to store `[KeyId]` related to each public key.

Maybe the easiest way to address this would be to modify the KeyInfo object as follows:

```
KeyInfo {
  private
  public
  primaryFingerprint: String
  primaryLongid: Strring
  allFingerprints: [String]
  allLongids: [String]
}
```

That would allow for all of the queries I mentioned.

Then do something very similar for the contacts / public keys table.

You can do this storage change PR without changing any of the logic, first. Just to have all of this information available in storage. Then in separate PR after that, we will discuss how to change app logic to take advantage of this and use fingerprints where possible.

**Comment by tomholub** [comment]
@Kharchevskyi do we let @ekievsky address this? Or do you worry it's too complex?

**Comment by Kharchevskyi** [comment]
I think it would be a bit complicated for me as well) 
So let's maybe I will pick this. What do you think @ekievsky?
